graphql_compiler/interpreter/typedefs.py

```diff
@@ -338,12 +338,140 @@ def project_neighbors(
+          it instead refers to "whatever value current_token points to at each generator iteration".
+        - Since the outer loop's iterations and the buggy generator's iterations may occur in
+          an arbitrarily-iterleaved order, the buggy generator's current_token value changes
+          unpredictably during its execution therefore causing the bug.
+        - Since Python generators are not started until their first result is requested, the above
+          buggy snippet can't even guarantee that the first use of current_token has the intended
+          value -- the outer loop might have iterated before the generator was first invoked.
+
+        This bug is avoided (as in the working example code) by calling a function and supplying
+        the value of current_token to it. This creates a new Python scope (a so-called "closure")
+        that remembers the value current_token had when the closure was created. A succinct,
+        internet-search-friendly way to describe the cause of this bug is to say that in Python,
+        generators are not closures.
+
+        ## Hints supplied to this function refer to neighboring vertices
+
+        Hint kwargs such as used_property_hints, filter_hints, and neighbor_hints in this function
+        describe the desired structure of the *neighboring* vertices that this function produces,
+        instead of to the vertices supplied via the data_contexts input. For example, consider
+        the following query:
+            {
+                Foo {
+                    out_Foo_Bar {
+                        name @output(out_name: "name")
+                    }
+                }
+            }
+        The project_neighbors() call corresponding to the out_Foo_Bar edge traversal would contain
+        used_property_hints=frozenset({"name"}). This is because project_neighbors() here would
+        be called DataContexts whose current_token would point to Foo vertices, and the neighboring
+        vertices (ones along the outbound Foo_Bar edge) have their "name" property queried.
+
+        Args:
+            data_contexts: iterable of DataContext objects which specify the DataTokens whose
+                           neighboring DataTokens need to be loaded.
+            current_type_name: name of the vertex type whose neighbors need to be loaded. Guaranteed
+                               to be the name of a type defined in the schema being queried.
+            edge_info: direction and name of the edge along which neighboring vertices need to be
+                       loaded. For example, in the query example above, this argument would be set
+                       to ("out", "Foo_Bar").
+            runtime_arg_hints: names and values of any runtime arguments provided to the query
+                               for use in filtering operations (e.g. "$arg_name").
+            used_property_hints: the property names of the neighboring vertices being loaded that
```

Comment:
```suggestion
            used_property_hints: property names of the neighboring vertices being loaded that are
```

Nit: argument descriptions don't have leading articles. Multi-line suggestions don't seem to be working for me but this suggestion would also mean removing the leading word "are" from the line below.

Reply:
Staged for commit.